### PR TITLE
when "dotnet test" runs test projects in parallel mode, json file writes invalid

### PR DESCRIPTION
**Pull request by loning**
dotnet test --no-build /p:CollectCoverage=true /p:CoverletOutputFormat='json%2copencover' /p:CoverletOutput="../results/coverage" /p:MergeWith="../results/coverage.json" /p:Exclude="[coverlet.*.tests?]*"

../results/coverage.json may faces concurrency problem, so when making the file share mode be none, it will solve the problem

### CODE REVIEW

File: src/coverlet.core/Coverage.cs [hunk 1]
```diff
@@ -32,7 +30,9 @@ public string Identifier
             get { return _identifier; }
         }
 
-        public Coverage(string module, string[] includeFilters, string[] includeDirectories, string[] excludeFilters, string[] excludedSourceFiles, string[] excludeAttributes, bool singleHit, string mergeWith, bool useSourceLink)
+        public Coverage(string module, string[] includeFilters, string[] includeDirectories, string[] excludeFilters,
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Coverage.cs]: Please could you revert formatting to old one here and on other places?

File: src/coverlet.core/Coverage.cs [hunk 2]
```diff
@@ -170,11 +183,22 @@ public CoverageResult GetCoverageResult()
...
+                var json = RetryHelper.Do(() =>
+                {
+                    using (var file = File.Open(_mergeWith, FileMode.Open, FileAccess.Read, FileShare.None))
+                    {
+                        using (var reader = new StreamReader(file))
+                        {
+                            return reader.ReadToEnd();
+                        }
+                    }
+                }, () => TimeSpan.FromMilliseconds(100), 5);
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Coverage.cs]: I'm not sure about max attempt, I would try for "more standard seconds" 10 or 30.

File: src/coverlet.core/Coverage.cs [hunk 3]
```diff
@@ -170,11 +183,22 @@ public CoverageResult GetCoverageResult()
...
             }
 
-            var coverageResult = new CoverageResult { Identifier = _identifier, Modules = modules, InstrumentedResults = _results };
+            var coverageResult = new CoverageResult
+                {Identifier = _identifier, Modules = modules, InstrumentedResults = _results};
 
             if (!string.IsNullOrEmpty(_mergeWith) && !string.IsNullOrWhiteSpace(_mergeWith) && File.Exists(_mergeWith))
             {
-                string json = File.ReadAllText(_mergeWith);
+                var json = RetryHelper.Do(() =>
```
Comment (MarcoRossignoli) [inline src/coverlet.core/Coverage.cs]: Please can you add some comment above like `// Possible concurrency access to merge file between write/read on parallel testing`  here and below?

**Comment by tonerdo** [comment]
@loning I'm not very clear the problem this PR is trying to solve

**Comment by loning** [comment]
I have lots of unit test project under one solution, and when we call "dotnet test any.sln", the unit test project will run in parallel at the same time, when we use /p:CoverletOutput="../results/coverage" /p:MergeWith="../results/coverage.json", may two project will write to ../results/coverage.json at the same time, and also when other project want to merge, it may read a file not completed.

so my solution is to make the file not shared, and when we cannot access a file, we wait 100ms, cause 100ms is enough for another read or write

https://travis-ci.org/AElfProject/AElf/builds/497464201?utm_source=github_status&utm_medium=notification

**Comment by loning** [comment]
I just found once error, and rerun the CI, disappeared. but I thought it's not hard to fix it

**Comment by MarcoRossignoli** [comment]
> and when we call "dotnet test any.sln",

@loning run coverlet solution wide is not a supported scenario for now https://github.com/tonerdo/coverlet/issues/350#issuecomment-466670698

**Comment by loning** [comment]
https://github.com/tonerdo/coverlet/pull/351#issuecomment-466682952

thanks，@MarcoRossignoli，but it seems worked well in our solution, we do not want to use shell to run test one by one

**Comment by MarcoRossignoli** [comment]
@loning for the moment I'll close this stale PR.
Even if you idea works well we should find a more robust way, we've an open issue on it https://github.com/tonerdo/coverlet/issues/357

**Pull request closed without merging**